### PR TITLE
Remove usage of self link for Kubernetes 1.20 support

### PR DESCRIPTION
**Pull request by rossf7**
Towards https://github.com/giantswarm/giantswarm/issues/16295

Removes usage of self link as its disabled by default in k8s 1.20.

Tested with https://github.com/giantswarm/chart-operator/pull/736

## Checklist

- [x] Update changelog in CHANGELOG.md.
- [ ] Update roadmap in ROADMAP.md.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 version: 2.1
 
 orbs:
-  architect: giantswarm/architect@0.17.0
+  architect: giantswarm/architect@dev:k8s-1-20
```
Comment (rossf7) [inline .circleci/config.yml:4]: Test with K8s 1.20.2 in the integration tests.

File: go.mod [hunk 2]
```diff
@@ -21,6 +21,7 @@ require (
 
 replace (
 	github.com/coreos/etcd => github.com/coreos/etcd v3.3.24+incompatible
+	github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
```
Comment (rossf7) [inline go.mod:24]: Fix nancy which flags vuln in 1.3.1.

File: pkg/controller/finalizer.go [hunk 3]
```diff
@@ -91,12 +92,12 @@ func (c *Controller) hasFinalizer(ctx context.Context, obj interface{}) (bool, e
 		return false, microerror.Mask(err)
 	}
 	finalizerName := getFinalizerName(c.name)
-	selfLink := accessor.GetSelfLink()
+	uid := string(accessor.GetUID())
```
Comment (rossf7) [inline pkg/controller/finalizer.go:95]: We now use UID as the cache key. It doesn't have a string function so just casting it.

https://github.com/kubernetes/apimachinery/blob/v0.20.5/pkg/types/uid.go#L19-L22

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -501,7 +501,7 @@ func (c *Controller) reconcile(ctx context.Context, req reconcile.Request, obj i
 		}
 	}
 
-	ctx = setLoggerCtxValue(ctx, loggerKeyObject, m.GetSelfLink())
+	ctx = setLoggerCtxValue(ctx, loggerKeyObject, fmt.Sprintf("%s/%s", m.GetNamespace(), m.GetName()))
```
Comment (rossf7) [inline pkg/controller/controller.go:504]: We lose the type information here which is a shame but its also less verbose in the logs.

**Review by rossf7 — COMMENTED** [review]
Testing on ginger.

**Review by fiunchinho — APPROVED** [review]
LGTM

**Review by kopiczko — APPROVED** [review]

**Pull request merged**